File: 1101-the-earliest-moment-when-everyone-become-friends/1101-the-earliest-moment-when-everyone-become-friends.cpp
class Solution {
public:
    
    int find(vector<int>&root,int x){
        
                while(x==root[x]){
                    
                        return x;
                }
                    
                return root[x]=find(root,root[x]);
    }   
    
    
    
    void unionfind(vector<int>&root,vector<int>&rank,int i,int j){
        
        
                    int rootx=find(root,i);
                    int rooty=find(root,j);
                    if(rootx!=rooty){
                        
                                if(rank[rootx]>rank[rooty]){
                                    
                                        rank[rootx]+=rank[rooty];
                                        root[rooty]=rootx;
                                        for(int i=0;i<root.size();i++){
                                            
                                                    if(root[i]==rooty) root[i]=rootx;
                                            
                                        }
                                    
                                }
                            
                                else{
                                    
                                        rank[rooty]+=rank[rootx];
                                        root[rootx]=rooty;
                                         for(int i=0;i<root.size();i++){
                                            
                                                    if(root[i]==rootx) root[i]=rooty;
                                            
                                        }
                                    
                                }
                        
                    }
            
                    
                    
        
        
        
    }
    
    
    int earliestAcq(vector<vector<int>>& logs, int n) {
     
                    sort(logs.begin(),logs.end());
                    vector<int>root(n,-1);
                    vector<int>rank(n,1);
        
                    for(int i=0;i<root.size();i++){
                        
                                root[i]=i;
                        
                    }
        
                   for(auto log:logs){
                       
                                unionfind(root,rank,log[1],log[2]);
                                set<int>s(root.begin(),root.end());
                                if(s.size()==1){
                                        return log[0];
                                }
                       
                   }
        
                    return -1;
                
        
    }
};

File: 1101-the-earliest-moment-when-everyone-become-friends/1101-the-earliest-moment-when-everyone-become-friends.py
class Solution:
    
    
    def find(self,parent,x):
        
            if parent[x]==x:
                    return x
                
            parent[x]=self.find(parent,parent[x])
            return parent[x]
    
    
    
    
    def helper(self,parent,i,j):
        
        parent_i=self.find(parent,i)
        parent_j=self.find(parent,j)
        
        
        parent[parent_j]=parent[parent_i]
        
        
    
    
    
    def earliestAcq(self, logs: List[List[int]], n: int) -> int:
            
            
            sorted_logs=sorted(logs,key=lambda logs:logs[0])
            
            
            parent=[i for i in range(n)]
            visited=set(i for i in range(n))
            
            for items in sorted_logs:
                
                self.helper(parent,items[1],items[2])
                for indx,item in enumerate(parent):
                    parent[indx]=self.find(parent,item)
                # print(parent,set(parent))
                if len(set(parent))==1:
                        return items[0]
                    
            return -1
            
            
            
            
        
            